include/lbann/layers/io/input/generic_input_layer.hpp

```diff
@@ -168,15 +169,20 @@ class generic_input_layer : public io_layer {
    * averaged across models. */
   void fp_setup_data(int mini_batch_size) override {
 
-    // Use the predetermined size of the mini-batch to set the current
-    // batch size for the neural network
+    // Determine model mini-batch size and effective mini-batch size
+    // Note: If inter-model communication is activated, the effective
+    // mini-batch is equal to the global mini-batch size.
     mini_batch_size = get_current_mini_batch_size();
-    this->m_model->set_current_mini_batch_size(mini_batch_size);
+    int effective_mini_batch_size = mini_batch_size;
+    for (auto&& cb : this->m_model->get_callbacks()) {
+      if (dynamic_cast<lbann_callback_imcomm*>(cb) != nullptr) {
+        effective_mini_batch_size = get_current_global_mini_batch_size();
```

Comment:
Since there _should_ be only one instance of `lbann_callback_imcomm` in any model, we can `break` after finding it.